### PR TITLE
Enforce TDD.

### PR DESCRIPTION
**Pull request by jctanner**
more context: https://github.com/ansible/galaxy_ng/pull/1236

[noissue]

Signed-off-by: James Tanner <tanner.jc@gmail.com>

### CODE REVIEW

File: .github/tdd.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    if not pr_branch:
+        pr_branch = json.loads(os.environ.get("GITHUB_CONTEXT"))["ref"]
+    print(f"PR-BRANCH: {pr_branch}")
+
+    # branch the PR wants to change
+    target_branch = os.environ.get("GITHUB_BASE_REF", "master")
+    print(f"TARGET-BRANCH: {target_branch}")
+    if target_branch != "master":
+        print("TDD is only enforced on the master branch")
+        sys.exit(0)
```
Comment (gerrod3) [inline .github/tdd.py:117]: Our default branch is named `main`.

File: .github/tdd.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    if not tests_found:
+        raise Exception("Tests should be added or modified with -every- PR!!!")
+
+
+def main():
+    logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
+
+    # these users should not be checked
+    skip_users = ["dependabot[bot]", "patchback[bot]"]
```
Comment (gerrod3) [inline .github/tdd.py:96]: Our automated CI (release process, plugin template updates) has [pulpbot](https://github.com/pulpbot) open PRs as well, think it should also be in this list.

File: .github/tdd.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        filenames = pid.stdout.decode("utf-8")
+        filenames = filenames.split("\n")
+        filenames = [x.strip() for x in filenames if x.strip()]
+    return filenames
+
+
+def verify_test_files_changed(changed_files):
+
+    # places where appplication source goes
+    app_paths = [f"{APP_NAME}/"]
```
Comment (gerrod3) [inline .github/tdd.py:53]: I think there are some files that shouldn't require test changes.

Also, what should happen in the case where a thing is deprecated in `pulpcore.plugin` and the change is simply switching over to the new one?

**Review by gerrod3 — COMMENTED** [review]
Overall I am in favor of this change, but I think there are some questions that need to be discussed first, mainly when this should be enforced.

**Comment by mdellweg** [comment]
Sorry, but i think my verdict here will be a clear "no".

This check looks whether some files have been changed. This does not tell me anything about whether the added/removed/changed tests are meaningful. I could change a whitespace to make this check happy. And it may make a reviewer just nod on the green check mark.
Requesting a changed / added test is really the responsibility of the (human) reviewer. Happy to put this on a checklist though.

BTW: Test-*driven*-development is still something different to me.

**Comment by awcrosby** [comment]
I’ll just add our rationale for adopting this in galaxy_ng… We took the mindset that almost all app changes should come with tests, this helped ensure tests were being thought about towards the start of the PR. It has really helped us increase the amount of tests written. Agreed test quality is still up to the author and reviewer.

We considered the ideal CI setup to be a tool to check that tests added in a PR relate to code touched in a PR. Or at least check coverage via a percentage of code that has been touched by tests, and for a given PR not let that percentage decrease. Even in both of those cases we would normally expect an edit to the dir where tests live. So due to complexity, we went with the simple check to start.

**Pull request closed without merging**